src/request.rs

```diff
@@ -236,438 +248,713 @@ impl<'a> Request<'a> {
-                se.filesystem.getxattr(
-                    self,
+                let response = se.filesystem.getxattr(
+                    self.meta,
                     self.request.nodeid().into(),
-                    x.name(),
-                    x.size_u32(),
-                    self.reply(),
+                    x.name().into(),
+                    x.size_u32()
                 );
+                match response {
+                    Ok(xattr)=> {
+                        self.replyhandler.xattr(xattr)
+                    }
+                    Err(err)=>{
+                        self.replyhandler.error(err)
+                    }
+                }
             }
             ll::Operation::ListXAttr(x) => {
-                se.filesystem
-                    .listxattr(self, self.request.nodeid().into(), x.size(), self.reply());
+                let response = se.filesystem.listxattr(
+                    self.meta,
+                    self.request.nodeid().into(),
+                    x.size()
+                );
+                match response {
+                    Ok(xattr)=> {
+                        self.replyhandler.xattr(xattr)
+                    }
+                    Err(err)=>{
+                        self.replyhandler.error(err)
+                    }
+                }
             }
             ll::Operation::RemoveXAttr(x) => {
-                se.filesystem.removexattr(
-                    self,
+                let response = se.filesystem.removexattr(
+                    self.meta,
                     self.request.nodeid().into(),
-                    x.name(),
-                    self.reply(),
+                    x.name().into()
                 );
+                match response {
+                    Ok(())=> {
+                        self.replyhandler.ok()
+                    }
+                    Err(err)=>{
+                        self.replyhandler.error(err)
+                    }
+                }
             }
             ll::Operation::Access(x) => {
-                se.filesystem
-                    .access(self, self.request.nodeid().into(), x.mask(), self.reply());
+                let response = se.filesystem.access(
+                    self.meta,
+                    self.request.nodeid().into(),
+                    x.mask()
+                );
+                match response {
+                    Ok(())=> {
+                        self.replyhandler.ok()
+                    }
+                    Err(err)=>{
+                        self.replyhandler.error(err)
+                    }
+                }
             }
             ll::Operation::Create(x) => {
-                se.filesystem.create(
-                    self,
+                let response = se.filesystem.create(
+                    self.meta,
                     self.request.nodeid().into(),
-                    x.name().as_ref(),
+                    x.name().into(),
```

Comment:
@rarensu here is where one of the copies happens that I am referring to

Reply:
I am beginning to understand. Path was constructed by borrowing from the underlying bytes, so there was no copy. Converting further to OsStr is zero copy, and converting even further to str could even be zero copy.  On the other hand, OsString is owned. There _must_ be a copy.

Reply:
What if I refactored the lower level request handling so that raw data is parsed into owned structs from the get-go? The upside of this is that the data stays owned at all times, so there is no need to copy. it's just move-move-move. The potential downside is that the data is no longer accessible to the session loop after it gets passed to the filesystem. But, is that an actual downside? Is there any reason that the session loop might need the raw request data after the filesystem takes it?